src/java/org/apache/cassandra/index/sai/utils/IndexFileUtils.java

```diff
@@ -59,25 +61,23 @@ public class IndexFileUtils
 
     private static final SequentialWriterOption writerOption = defaultWriterOption;
 
+    /**
+     * Remembers checksums of files so we don't have to recompute them from the beginning of the file whenever appending
+     * to a file. Keeps checksums with respective file lengths so we can detect some file changes like truncation
+     * or external appends which could omit this code, so this is still safe.
+     * Index files are append-only (immutable), so length checking is enough.
+     */
+    private static final Cache<String, FileChecksum> checksumCache = Caffeine.newBuilder().maximumSize(4096).build();
```

Comment:
I am not sure that this is a good idea, because we don't have control about concurrent access.

Let's say there is some problem and 2 IndexOutputWriters start on the same path, for instance because an operation started and it is very slow, then something triggered a new index build operations, then 2 threads work on the same checksum instance.

I would change this to ensure that a FileChecksum instance is owned by only an  IndexOutputWriter

Reply:
Cache entries are per file. 
If there are things appending to the same file concurrently, we've got a bigger problem.

Reply:
But ok, I agree let's harden it more.

Reply:
This is hardened now, the FileChecksum instance is locked to one user at a time.